`reference.relativeDirectory` validation

austeja-bentley opened the pull request:
In this PR:
- Added validation for object reference's `relativeDirectory` property. Invalid `relativeDirectory` values are:
  - values that contain backslashes
  - values that have a trailing slash
  - value that have a slash at the beginning of a string
- Added unit tests for the `assertRelativeDirectory` method.
- Added unit tests to test that storage class methods assert `relativeDirectory` where applicable.

Animagne commented on tests/unit/src/BindingsTests.ts:
```diff
@@ -0,0 +1,74 @@
+import { ServerStorageDependency } from "@itwin/object-storage-core";
+
+export interface DependencyBindingsTestCase {
+  symbolUnderTestName: string;
```
This is an example, where modifier ends up in the middle and I'm not really sure what this really means anymore. Should this be something like `testedBindingsSymbol`, `testedBindingsName` or `testedClassIdentifier`? Symbol and string are very similar primative data types and `stringUnderTestName` would definitely sound wrong.

austeja-bentley replied:
This was also addressed in the previously merged PR: https://github.com/iTwin/object-storage/pull/47

austeja-bentley commented on storage/core/src/frontend/Helpers.ts:
```diff
@@ -147,6 +147,25 @@ export function metadataToHeaders(
+  if (!relativeDirectory) return;
+
+  const backslash = "\\";
+  if (relativeDirectory.includes(backslash) || relativeDirectory === "")
```
Removed: https://github.com/iTwin/object-storage/pull/46/commits/362e8138c74cdeabaa05afc869f6fedc248e08df

austeja-bentley commented on tests/unit/src/CommonTests.ts:
```diff
@@ -2,73 +2,17 @@
+use(chaiAsPromised);
+
+export async function testRelativeDirectoryValidation(
+  functionUnderTest: () => Promise<unknown>
```
Removed all `underTest` occurrences in previous PR, fixed the current changes accordingly.

austeja-bentley commented:
@Animagne 
as per our internal discussion I have reworked the structure on how the test code is shared. I have added `@itwin/object-storage-tests-backend-unit` package that contains:
1. common unit tests that test `ClientStorage` and `ServerStorage`. For now we only check the `relativeDirectory` validation. These tests are configured and run the same way our integration tests are - there's an exposed runner class (`StorageUnitTests`) that accepts binding class constructors, calls `.register`, retrieves storage classes and runs the tests against them. Bindings that are passed to unit tests are reconfigured to use mocks for internal dependencies so that they do send `http` requests.
2. `shared` folder which holds shared test utils that are imported by unit tests in specific storage provider packages.

For now I have ignored any `FrontendStorage` related tests as the next task is to configure them to run on `cypress`.

austeja-bentley commented on storage/minio/src/test/unit/backend/RunCommonUnitTests.ts:
```diff
@@ -0,0 +1,69 @@
+tests.start().catch((err) => {
+  process.exitCode = 1;
+  throw err;
+});
```
I think if there was no empty line github would show this - 
![image](https://user-images.githubusercontent.com/70565417/167999415-73b0d5bd-051b-4b43-a70c-d3a1babc78bb.png)
Since there is no such symbol the empty line is there, it is also visible in locally checked-out repo. I would personally prefer if there was no trickery in regards to displaying empty lines at the end of file..

Empty lines at the end of `.ts` files are enforced by `prettier` ESLint plugin, linting does result in an error if no newline is present.

austeja-bentley commented on cloud-agnostic/core/package.json:
```diff
@@ -52,4 +53,4 @@
     "node": ">=12.20 <17.0.0"
   }
-}
+}
```
Fixed: https://github.com/iTwin/object-storage/pull/46/commits/0056c0ad35afb4fcf11b371988d1feaeeb21ecb1

austeja-bentley commented on storage/core/package.json:
```diff
@@ -54,4 +55,4 @@
     "node": ">=12.20 <17.0.0"
   }
-}
+}
```
Fixed: https://github.com/iTwin/object-storage/pull/46/commits/0056c0ad35afb4fcf11b371988d1feaeeb21ecb1
